tests/test_frontend_python_code_check.py

```diff
@@ -0,0 +1,140 @@
+import ast
+import pytest
+from xdsl.frontend.exception import CodeGenerationException
+
+from xdsl.frontend.python_code_check import Constant, ConstantVisitor
+
+def test_const_correctly_evaluated():
+    visitor = ConstantVisitor()
+    src = \
+"""
+a: Const[i32] = 2 ** 5
+b: Const[i32] = len([1, 2, 3])
+"""
+    visitor.visit(ast.parse(src))
+
+    assert "a" in visitor.constants
+    assert not visitor.constants["a"].shadowed
+    assert visitor.constants["a"].value == 32
+
+    assert "b" in visitor.constants
+    assert not visitor.constants["b"].shadowed
+    assert visitor.constants["b"].value == 3
+
+def test_can_assign_to_shadowed_constants_I():
```

Comment:
Here the variables get shadowed, right? Could you assert that they actually are and that their value is right?